staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go

```diff
@@ -0,0 +1,453 @@
+)
+
+const (
+	azureTimeFormat = "\"2006-01-02 15:04:05.99999\""
+	azureAuthority  = "https://login.microsoftonline.com"
+	uuidFormat      = "[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[8|9|aA|bB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}"
+	azureTokenKey   = "azureTokenKey"
+
+	cfgClientID     = "client-id"
+	cfgTenantID     = "tenant-id"
+	cfgAudience     = "audience"
+	cfgAccessToken  = "access-token"
+	cfgRefreshToken = "refresh-token"
+	cfgTokenType    = "token-type"
+	cfgExpiresIn    = "expires-in"
+	cfgExpiresOn    = "expires-on"
+)
+
+func init() {
+	if err := restclient.RegisterAuthProviderPlugin("azure", newAzureAuthProvider); err != nil {
+		glog.Fatalf("Failed to register azure auth plugin: %v", err)
+	}
+}
+
+var cache = newAzureTokenCache()
+
+type azureTokenCache struct {
+	lock  sync.Mutex
+	cache map[string]*azureToken
+}
+
+func newAzureTokenCache() *azureTokenCache {
+	return &azureTokenCache{cache: make(map[string]*azureToken)}
+}
+
+func (c *azureTokenCache) getToken(tokenKey string) *azureToken {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.cache[tokenKey]
+}
+
+func (c *azureTokenCache) setToken(tokenKey string, token *azureToken) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.cache[tokenKey] = token
+}
+
+func newAzureAuthProvider(_ string, cfg map[string]string, persister restclient.AuthProviderConfigPersister) (restclient.AuthProvider, error) {
+	var ts tokenSource
+
+	clientID := cfg[cfgClientID]
+	tenantID := cfg[cfgTenantID]
+	audience := cfg[cfgAudience]
+	if clientID != "" && tenantID != "" && audience != "" {
+		ts = newAzureTokenSourceDeviceCode(clientID, tenantID, audience)
+	} else {
+		azFilename, err := azureTokensFile()
+		if err != nil {
+			return nil, err
+		}
+		ts, err = newAzureTokenSourceFromFile(azFilename)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	cacheSource := newAzureTokenSourceFromCache(ts, cache, cfg, persister)
+
+	return &azureAuthProvider{
+		tokenSource: cacheSource,
+	}, nil
+}
+
+type azureAuthProvider struct {
+	tokenSource tokenSource
+}
+
+func (p *azureAuthProvider) Login() error {
+	return errors.New("not yet implemented")
+}
+
+func (p *azureAuthProvider) WrapTransport(rt http.RoundTripper) http.RoundTripper {
+	return &azureRoundTripper{
+		tokenSource:  p.tokenSource,
+		roundTripper: rt,
+	}
+}
+
+type azureRoundTripper struct {
+	tokenSource  tokenSource
+	roundTripper http.RoundTripper
+}
+
+func (r *azureRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	token, err := r.tokenSource.Token()
+	if err != nil {
+		glog.Errorf("Failed to acquire a token: %v", err)
+		return nil, err
+	}
+
+	// clone the request in order to avoid modifying the headers of the original request
+	req2 := new(http.Request)
+	*req2 = *req
+	req2.Header = make(http.Header, len(req.Header))
+	for k, s := range req.Header {
+		req2.Header[k] = append([]string(nil), s...)
+	}
+
+	req2.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.token.AccessToken))
```

Comment:
@cosmincojocar the audience shouldn't be the API server, it should be whatever the client_id of the client that initiated the login flow to request an ID token. That's specifically called out in the docs. The API server is NOT a client. It's set up to only allow ID tokens issues for a single client.

As for that snippet, it's just implementing the spec[0]

> 3. The Client MUST validate that the aud (audience) Claim contains its client_id value registered at the Issuer identified by the iss (issuer) Claim as an audience. The aud (audience) Claim MAY contain an array with more than one element. The ID Token MUST be rejected if the ID Token does not list the Client as a valid audience, or if it contains additional audiences not trusted by the Client.

The issue seems to be the `aud` embedded in your access token, which is `https://management.core.windows.net/` rather then your client_id.

[0] https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation